File: Data Structures & Algorithms/maximum-frequency-stack/submission-3.py
class FreqStack:

    def __init__(self):
        self.items = defaultdict(list)
        self.count = defaultdict(int)
        self.max_count = 0

    def push(self, val: int) -> None:
        self.count[val] += 1
        self.items[self.count[val]].append(val)
        self.max_count = max(self.max_count, self.count[val])

    def pop(self) -> int:
        mc = self.max_count
        val = self.items[mc].pop()
        if not self.items[mc]: self.max_count -= 1
        self.count[val] -= 1
        return val
        

        


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()

File: Data Structures & Algorithms/implement-queue-using-stacks/submission-3.cpp
class MyQueue {
private:
    stack<int> main_q;
    stack<int> buffer; 

    void populate_buffer() {
        while (!main_q.empty()) {
            buffer.push(main_q.top());
            main_q.pop();
        }
    }

public:
    void push(int x) {
        main_q.push(x);
    }
    
    int pop() {
        if (buffer.empty())
            populate_buffer();
        int val = buffer.top();
        buffer.pop();
        return val;
    }
    
    int peek() {
        if (buffer.empty())
            populate_buffer();
        return buffer.top();
    }
    
    bool empty() {
        return (main_q.empty() && buffer.empty());
    }


};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */